### PR TITLE
(CDAP-15046) Use deployment watch to enforce correct state change

### PR DESCRIPTION
**Pull request by chtyim**
- Use deployment watch to implement LiveInfo cache
- Use deployment watch to enforce max application start timeout
- Use deployment watch to transit controller state

### CODE REVIEW

**Comment by chtyim** [comment]
Build passed https://builds.cask.co/browse/CDAP-DUT6924-2

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/discovery/KubeDiscoveryService.java [hunk 1]
```diff
@@ -352,168 +343,68 @@ private void closeQuietly(@Nullable AutoCloseable closeable) {
...
-        CoreV1Api coreApi = getCoreApi();
-        Call call = coreApi.listNamespacedServiceCall(namespace, null, null, null, null, labelSelector,
-                                                      null, null, null, true, null, null);
-        this.watch = watch = Watch.createWatch(KubeDiscoveryService.this.coreApi.getApiClient(), call,
-                                               new TypeToken<Watch.Response<V1Service>>() { }.getType());
-        return watch;
-      }
+    @Override
+    public void resourceModified(V1Service service) {
+      // Treat modify the same as add since both would contains the complete list of
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/discovery/KubeDiscoveryService.java]: nit: contain

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  }
+
+  /**
+   * Returns a {@link Call} object for the watcher to use.
+   *
+   * @param namespace namespace for the call to operate on
+   * @param labelSelector the label selector to use for selecting resource to watch
+   * @return a {@link Call} for {@link Watch} to use
+   */
+  protected abstract Call createCall(String namespace, String labelSelector) throws IOException, ApiException;
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java]: `labelSelector` should be marked as nullable.

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    synchronized (this) {
+      // Need to check here for the case where the thread that calls the close() method set the flag to true,
+      // then this thread grab the lock and update the watch field.
+      if (stopped) {
+        return null;
+      }
+
+      // There is only single thread (the run thread) that will call this method,
+      // hence if the watch was null outside of this sync block, it will stay as null here.
+
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java]: nit: extra line

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    int failureCount = 0;
+    while (!stopped) {
+      try {
+        Watch<T> watch = getWatch();
+
+        // If we can create the watch, reset the failure count.
+        failureCount = 0;
+
+        // The watch can only be null if the watcher thread is stopping
+        if (watch == null) {
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java:131]: should we move this closer to `watch` assignment?

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java [hunk 5]
```diff
@@ -161,7 +161,7 @@ public final void run() {
 
         // We just retry on any form of exceptions
         Throwable cause = e.getCause();
-        if (cause instanceof IOException || cause instanceof IllegalStateException) {
+        if (cause instanceof IOException || e instanceof IllegalStateException) {
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java:165]: why is this change needed?

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillRunnerService.java [hunk 6]
```diff
@@ -187,40 +187,179 @@ public Cancellable setSecureStoreRenewer(SecureStoreRenewer renewer, long initia
...
+                                                TimeUnit timeoutUnit, KubeTwillController controller) {
+    String runId = controller.getRunId().getId();
+
+    LOG.debug("Monitoring application {} with run {} starts in {} {}",
+              liveInfo.getApplicationName(), runId, timeout, timeoutUnit);
+
+    // Schedule to terminate the controller in the timeout time.
+    Future<?> terminationFuture = monitorScheduler.schedule(controller::terminate, timeout, timeoutUnit);
+
+    // This queue is for transferring the cancel watch to the
```
Comment (CuriousVini) [inline cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillRunnerService.java]: incomplete comment?

**Review by rohitsinha54 — APPROVED** [review]

**Review by CuriousVini — APPROVED** [review]
Few minor comments. Rest lgtm 👍

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/discovery/KubeDiscoveryService.java (hunk 1 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/discovery/KubeDiscoveryService.java]: fixed

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java (hunk 2 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java]: fixed

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java (hunk 3 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java]: fixed

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java (hunk 4 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java:131]: Done.

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java (hunk 5 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/common/AbstractWatcherThread.java:165]: Because the K8s API don't wrap IllegalStateException.

File: cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillRunnerService.java (hunk 6 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillRunnerService.java]: fixed
